### ISSUE TITLE
Model organisms: composable framework for synthetic LLM experiments

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/predicting-zygosity#418]
# Model organisms: a composable framework for synthetic LLM experiments

## Summary

Build a unified framework for generating synthetic datasets and evaluating fine-tuned models on tasks with known properties. These "model organisms" serve dual purposes: verifying that the fine-tuning and evaluation pipelines work correctly, and enabling rapid scientific experimentation on questions like learning speed, memorization vs. generalization, and out-of-distribution transfer.

The framework replaces the existing ad-hoc sanity checks (`bit_sequences/`, `predictable_or_not/`) with a small set of composable building blocks that can be recombined to produce many different experiments.

## Why "model organisms"

In biology, model organisms (fruit flies, C. elegans, mice) are small, well-understood systems used for rapid experimentation. Insights gained from them transfer to more complex systems. These synthetic tasks play the same role for LLM fine-tuning: they're cheap to run, have known expected outcomes, and the lessons learned inform experiments on real data.

## Framework design

An experiment is specified by choosing from three dimensions:

### 1. Input type

The input is a sequence of tokens drawn from one of these domains:

| Input type | Alphabet | Example |
|------------|----------|---------|
| `bits` | {0, 1} | `0 1 1 0 1` |
| `digits` | {0, 1, ..., 9} | `3 7 2 9 4` |
| `letters` | {a, b, ..., z} | `m i c k y` |

Each input type has a parameter **k** (sequence length).

### 2. Output rule

A function applied to the input sequence to produce a single-token output. Rules are either **universal** (work on any input type) or **domain-specific**.

**Universal rules:**

| Rule | Description | Example (input: `3 7 2 9 4`) |
|------|-------------|------|
| `length` | Number of elements in the sequence | `5` |
| `first` | First element | `3` |
| `last` | Last element | `4` |
| `nth(x)` | Element at position x | `2` (x=3) |
| `constant` | Always output the same token (default: 42) | `42` |
| `coin(p)` | Random binary label, probability p of outputting 1, fixed by seed. No relationship to input. | `0` or `1` |

**Domain-specific rules:**

| Rule | Input types | Description |
|------|-------------|-------------|
| `parity` | bits | 1 if odd number of 1s, else 0 |
| `majority` | bits | 1 if more 1s than 0s, else 0 |
| `min` | digits, letters | Smallest element |
| `max` | digits, letters | Largest element |

Notes:
- `constant` is configurable; default is 42.
- `coin(p)` labels are assigned randomly at generation time and fixed by seed. In the memorization setting, these are learnable (just memorize the mapping). In the generalization setting, they are the negative control — if the model beats chance on held-out data, something is wrong (likely data leakage).
- Not every input type × rule combination needs to exist. The framework should make it easy to add combinations without requiring completeness.
- Any deterministic rule can optionally have a noise parameter (probability of label flip) to create signal + noise conditions.

### 3. Format

How the sequence is rendered as text, which affects tokenization:

| Format | Example (bits, k=5) |
|--------|------|
| `dense` | `01101` |
| `spaced` | `0 1 1 0 1` |
| `comma` | `0,1,1,0,1` |
| `tab` | `0\t1\t1\t0\t1` |
| `pipe` | `0\|1\|1\|0\|1` |

Hypothesis: tokenizer-friendly formats (where each element is a separate token) learn faster, but both eventually learn. This is testable within the framework by varying format while holding everything else constant.

### 4. Experimental design

Three designs, each answering different questions:

**Memorization**: Eval set = training set. Tests whether the pipeline can teach the model anything at all. Expected accuracy is near-perfect for all rules given enough training, including `coin(p)`. If accuracy is low, the pipeline is broken.

**Generalization in-distribution**: Eval set is new samples from the same data-generating process (e.g., 80/20 split of sequences at the same k). Tests whether the model learns the rule, not just the training examples. Expected accuracy depends on the rule:
- Deterministic rules (parity, first, max, etc.): high if the model can learn the rule; ~chance if it can't (this is an empirical question for some rules)
- `constant`: near-perfect (the rule trivially generalizes — positive control)
- `coin(p)`: ~50% (no rule to generalize — negative control)

**Generalization out-of-distribution**: Eval set comes from a different distribution than training. The user specifies one in-distribution training set and one or more OOD test sets. Three OOD axes:
- **Sequence length**: train on k=5, test on k=7
- **Format**: train on dense, test on spaced
- **Input domain**: train on bits, test on digits (universal rules only)

Cross-domain OOD (changing input type) only works for universal rules. The framework should warn or error on invalid combinations (e.g., testing `parity` on digits).

## Parameters

A single generated dataset is fully specified by these parameters:

| Parameter | Description | Example |
|-----------|-------------|---------|
| `input_type` | bits, digits, or letters | `bits` |
| `rule` | Output function | `parity` |
| `k` | Sequence length | `8` |
| `format` | Text rendering | `spaced` |
| `N` | Dataset size | `1000` |
| `seed` | Random seed | `1729` |
| `p` | For `coin(p)`: probability of 1. For noise on other rules: flip probability. | `0.5` |
| `constant_value` | For `constant` rule | `42` |

An experiment additionally specifies:

| Parameter | Description | Example |
|-----------|-------------|---------|
| `design` | memorization, in_distribution, ood | `in_distribution` |
| `split` | Train/test ratio (in_distribution only) | `0.8` |
| `ood_tests` | List of test dataset specs, each with the parameters above (ood only) | one per OOD axis |

## Scientific questions this enables

By varying parameters while holding others fixed, the framework supports experiments like:

- **Does the pipeline work?** Memorization on any deterministic rule — accuracy should approach 100%.
- **Can the model learn rule X?** In-distribution generalization — compare accuracy to the `coin(p)` negative control and `constant` positive control.
- **How quickly does it learn?** Evaluate at each epoch checkpoint. Compare learning curves across rules of different complexity.
- **How data-efficient is learning?** Same rule, vary N. At what point does accuracy saturate?
- **Does format affect learning?** Same task, vary format (dense vs. spaced vs. comma). Measure learning speed.
- **Does learning transfer across sequence lengths?** Train on k=5, test on k=7 and k=10.
- **Does learning transfer across input domains?** Train `first` on bits, test `first` on digits.
- **Can the model distinguish signal from noise?** Compare in-distribution generalization on a deterministic rule vs. `coin(0.5)`. The model should succeed on one and fail on the other.

## Desiderata

1. **Composability without completeness.** Input types and rules combine freely, but the framework does not require filling every cell of the matrix. Easy to add new combinations.
2. **Single data format.** All generated datasets produce the same JSON structure regardless of input type or rule. One parameterized eval task replaces the three existing near-identical eval tasks.
3. **Reproducibility.** Same parameters + same seed = identical dataset. Seed is a required parameter.
4. **Self-documenting datasets.** Each generated dataset includes metadata about its data-generating process (input type, rule, parameters, seed) so experiments are auditable without external records.
5. **Known expected outcomes where possible.** For each combination, the framework states what accuracy to expect — or explicitly flags it as an open question.
6. **Cheap to run.** Small datasets, short training runs. The framework supports exhaustive enumeration for small k and sampling for larger k.

## Integration with cruijff_kit workflow

Model organism experiments should use the existing experiment workflow — **not** a parallel system. The `design-experiment` skill should accept model organism specifications (e.g., "I want a memorization experiment with bit parity at k=8") and handle data generation as part of the design process, producing an `experiment_summary.yaml`. From there, `scaffold-experiment` and `run-experiment` work as usual.

This is a single point of entry, consistent with how Sarah's `tabular_to_text` branch handles data generation within `design-experiment`.

### Expressing OOD experiments in experiment_summary.yaml

The existing schema already supports "fine-tune once, evaluate on many datasets" via per-task dataset fields in `evaluation.tasks[]`. No schema changes are needed. A model organism OOD experiment translates to:

```yaml
data:
  training:
    path: data/green/model_organisms/bits_k8_parity_dense.json

evaluation:
  tasks:
    - name: in_distribution
      script: sanity_checks/model_organisms/inspect_task_model_organisms.py
      dataset: same  # uses training data's validation split
      vis_label: in_dist
    - name: ood_length
      script: sanity_checks/model_organisms/inspect_task_model_organisms.py
      dataset: data/green/model_organisms/bits_k10_parity_dense.json
      vis_label: ood_k10
    - name: ood_format
      script: sanity_checks/model_organisms/inspect_task_model_organisms.py
      dataset: data/green/model_organisms/bits_k8_parity_spaced.json
      vis_label: ood_spaced
```

All tasks use the same parameterized inspect task (the unified eval task), distinguished by `vis_label` for analysis. The evaluation matrix should run each task at every epoch checkpoint (evaluation is cheap relative to fine-tuning, and per-epoch accuracy curves reveal learning dynamics that a final-only evaluation would miss).

No experiment_summary.yaml shorthand for generating OOD variants. That file is already complex and should stay as simple as possible. `design-experiment` writes explicit task entries.

### Generated datasets

Datasets are generated on demand, not committed to git. Reproducibility is guaranteed by the seed and data-generating-process metadata embedded in each generated file. Users regenerate datasets when they need them.

## Relationship to existing code

**Eval task:** The three existing eval tasks (`inspect_task_bit_sequences.py`, `inspect_task_predictable_or_not.py`, `inspect_task_capitalization.py`) are nearly identical. The unified eval task is a parameterized version of any one of them. Heavy reuse (~80% existing code).

**Generator:** The `bit_sequences/generate_input_data.py` is the closest ancestor. Its structure (enumeration, splitting, deduplication, seed handling, JSON output) becomes the skeleton. Substantial refactoring to support the input type × rule × format matrix. Medium reuse (~40% existing code, ~60% new).

**Individual rules:** Each is 1-5 lines of logic. `parity` already exists; the rest are new but trivial.

**predictable_or_not:** The `up` (unpredictable input, predictable output) and `uu` (both unpredictable) scenarios map onto the framework as `constant` and `coin` rules respectively. The `pp` and `pu` scenarios (which use sequential integers and a "next number" rule) do not have a natural home in this framework — that's a different kind of task and is not being migrated.

## Implementation order

One issue, built bottom-up so each layer is testable before the next:

1. **Generator framework + bits + core rules** (parity, first, constant, coin) + memorization and in-distribution designs
2. **Remaining rules and input types** (digits, letters, majority, min, max, last, nth, length) — registering new functions into the existing framework
3. **OOD support and format variations**
4. **Unified eval task** replacing the three existing tasks
5. **design-experiment integration**

## What this replaces

The framework lives at `sanity_checks/model_organisms/` (the eventual `projects/` restructuring is a separate issue).

Once complete, this framework supersedes:
- `sanity_checks/bit_sequences/` (generator + eval task)
- `sanity_checks/predictable_or_not/` (generator + eval task)
- `sanity_checks/bit_parity_test/issue.md` (spec only, never implemented)
- The bit_parity datasets in `data/green/bit_parity_{5,7,9}bit/`

The `experiments/capitalization/` experiment is a real-data project and remains separate, though its eval task could optionally be migrated to the unified eval task.

## Related issues

- #401 — Create a test where we know the right answer (this issue generalizes and supersedes the discussion there)
- Restructuring `experiments/` and `sanity_checks/` into `projects/` (separate issue, not yet filed)

**Comment by niznik-dev** [comment]
Framework shipped in PR #423 — all 5 implementation-order steps landed (generator, rules, inputs, formats, unified eval task, design-experiment integration).

Follow-ups tracked on their own issues:
- #434 — remove pre-framework scripts that this framework supersedes
- #436 — re-add multi-dataset support to data.data_generation
- #435 — inspect-ai list[SampleScore] migration (tangential, from the calibration work in chunk 4)

Future rule/input additions will be filed as new enhancement issues rather than reopening this umbrella.

—MxC
